fix: add DynamicPromptsFunction and DynamicActionsFunction to type definitions

armand1m opened the pull request:
When the PR #37 got merged, Dynamic Prompt functions got implemented and supported, but types ain't got updated.

This also adds a `PromptAnswers` type, which is just an alias to `Record<string, any>`

This PR fixes #162, fixes #142 and fixes #140

armand1m commented:
it will need some changes, it looks like the return should not be an array and there is actually an inquirer instance that gets passed as an argument to the prompts call

crutchcorn commented:
I'm working on a PR to introduce TS typing tests, which may impact this PR. Just a heads up

armand1m commented:
@crutchcorn noice, thanks for the heads up! I will update your tests once I get this done

crutchcorn commented:
Alright, so I've moved the `index.d.ts` file to `types` and added a `tests.ts` file. Can you update this PR with the new file location but also write some typing tests as well?

armand1m commented:
@crutchcorn sure thing!

armand1m commented:
Fixed merge conflicts, updated the new types definition and bumped @types/inquirer to 6.5.0 to be able to incorporate #167 and avoid another merge conflict. Still need to add the tests to the type definitions.

I also updated `inquirer` to 7.1.0 as a convenience, tests still pass so I believe it is safe. @crutchcorn if you don't feel like bumping this, please let me know so I can revert 🙂

armand1m commented on types/index.d.ts:
```diff
@@ -58,10 +58,27 @@ interface PlopActionHooks {
+	| inquirer.RawListQuestion[]
+	| inquirer.PasswordQuestion[]
+	| inquirer.NumberQuestion[]
+	| inquirer.InputQuestion[];
```
one thought that comes to my mind here is that this limits the prompts array to always be of a specific type of question.

I'm going to change this to something like this so you can have different types of questions within the same array 

```ts
type PromptQuestion = inquirer.Question
  | /** other types */

type Prompts = DynamicPromptsFunction | PromptQuestion[];
```

armand1m replied:
updated this :)

armand1m commented:
@crutchcorn added some tests and also had to fix some types because they were not matching the current implementation. let me know what you think

crutchcorn commented:
This looks perfect! Thank you so much!
